### PR TITLE
Add notifications inline action for share an own published post

### LINKED ISSUE #22473: 🍎 Share an own published post
The users should be able to share their own posts. This should apply on notifications for a post like or a scheduled post that went live (design yWt5gg3nWORhu079Qfv3mS-fi-51_1048)

An event should also be tracked. See: pc8HXX-1ua-p2

### PR DESCRIPTION
**Pull request by salimbraksa**
Fixes #22473

## Description

This PR adds an inline action to share a post.

## Test Instructions

* Log in to the Jetpack app (with an account containing various notifications)
* Tap the notifications tab
* Scroll to a notification for a post like
* Expect to see a share icon
* Tap the share icon
* Expect the share activity to launch
* Expect this event to be tracked `notifications_inline_action_tapped <inline_action: share_post>`

Test for regression:

* Tap the other part of the notification
* Expect the notification detail view to be presented (as usual)

## Regression Notes
1. Potential unintended areas of impact
Notification cell configuration logic was refactored, so double check the notification content is correct.

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Manual testing.

3. What automated tests I added (or what prevented me from doing so)
None.

## PR submission checklist:

- [x] have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

Testing checklist:
- [ ] WordPress.com sites and self-hosted Jetpack sites.
- [ ] Portrait and landscape orientations.
- [ ] Light and dark modes.
- [ ] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] VoiceOver.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [ ] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [ ] iPhone and iPad. 
- [ ] Multi-tasking: Split view and Slide over. (iPad)

### CODE REVIEW

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 500 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationsViewController/NotificationsViewController.swift [hunk 1]
```diff
@@ -381,6 +353,64 @@ class NotificationsViewController: UIViewController, UIViewControllerRestoration
...
+            .init(
+                title: title,
+                description: description,
+                shouldShowIndicator: !notification.read,
+                avatarStyle: avatarStyle,
+                inlineAction: inlineAction
+            )
+        )
+        return .init(style: style)
+    }
```
Comment (salimbraksa) [inline WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationsViewController/NotificationsViewController.swift:390]: No major changes here, I've simply extracted this logic from `cellForRowAtIndexPath` method into this function.

File: WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationsViewController/NotificationsViewController.swift [hunk 2]
```diff
@@ -381,6 +353,64 @@ class NotificationsViewController: UIViewController, UIViewControllerRestoration
...
+            )
+        case .like, .reblog:
+            return .init(
+                icon: Image.DS.icon(named: .blockShare),
+                action: { [weak self] in self?.viewModel.sharePostActionTapped(with: notification, at: indexPath) }
+            )
+        default:
+            return nil
+        }
+    }
```
Comment (salimbraksa) [inline WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationsViewController/NotificationsViewController.swift:409]: The `like` comment and post action is not part of this PR.

File: WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationsViewController/NotificationsViewController.swift [hunk 3]
```diff
@@ -338,47 +338,94 @@ class NotificationsViewController: UIViewController, UIViewControllerRestoration
...
-        cell.accessibilityHint = Self.accessibilityHint(for: note)
-        cell.host(NotificationsTableViewCellContent(style: style), parent: self)
-
-        return cell
+        // Handle the regular case
+        let title: AttributedString? = {
+            guard let attributedSubject = notification.renderSubject() else {
+                return nil
+            }
+            return AttributedString(attributedSubject)
```
Comment (alpavanoglu) [inline WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationsViewController/NotificationsViewController.swift:375]: No need to change anything as it is my personal taste:
Since there's only 1 line after guard here, I think the former if-else is an easier read than this.

File: WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationsViewController/NotificationsViewModel.swift [hunk 4]
```diff
@@ -1,17 +1,39 @@
+import Foundation
+import AutomatticTracks
+
 final class NotificationsViewModel {
```
Comment (justtwago) [inline WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationsViewController/NotificationsViewModel.swift:4]: Would it be possible to add unit tests to this class?

**Review by justtwago — COMMENTED** [review]
Thanks for this feature, Salim! 🎸 
Tested it on my physical device and it works as described.
I also see the event: `🔵 Tracked: notifications_inline_action_tapped <inline_action: share_post>`

Just left a comment about missing unit tests.

File: WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationsViewController/NotificationsViewModel.swift [hunk 5]
```diff
@@ -73,4 +95,34 @@ final class NotificationsViewModel {
...
+        self.trackInlineActionTapped(action: .sharePost)
+        return content
+    }
+}
+
+// MARK: - Analytics Tracking
+
+private extension NotificationsViewModel {
+
+    func trackInlineActionTapped(action: InlineAction) {
```
Comment (justtwago) [inline WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationsViewController/NotificationsViewModel.swift:132]: Shouldn't it be a private func?

File: WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationsViewController/NotificationsViewModel.swift [hunk 6]
```diff
@@ -73,4 +95,34 @@ final class NotificationsViewModel {
...
+        self.analyticsTracker.track(.notificationsInlineActionTapped, properties: ["inline_action": action.rawValue])
```
Comment (justtwago) [inline WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationsViewController/NotificationsViewModel.swift:133]: Is it recommended to extract such keys as `"inline_action"` to constants?

**Review by justtwago — COMMENTED** [review]
And a couple of minor comments 🙇

File: WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationsViewController/NotificationsViewModel.swift (hunk 5 above)
Comment (salimbraksa) [reply WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationsViewController/NotificationsViewModel.swift:132]: AFAIK when an extension is `private`, all methods within are also `private`

File: WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationsViewController/NotificationsViewModel.swift (hunk 6 above)
Comment (salimbraksa) [reply WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationsViewController/NotificationsViewModel.swift:133]: Since the string `inline_action` is specific to this tracking event and is not used elsewhere nor anticipated to be used again, defining it as a constant may not be necessary. This keeps our code simple and avoids an overload of single-use constants.

**Review by osullivanchris — APPROVED** [review]
This is working as expected for me. One note. The share icon is centred vertically with the user avatar, which is as expected. In this build I also see star icons to like post, those are not centered vertically as expected.

**Review by hassaanelgarem — COMMENTED** [review]
The code looks good to me and it works as expected. One bit I think we can improve is the message we share along with the post URL. Right now we only share the URL.

I implemented my suggestion for adding a message and pushed it [here](https://github.com/wordpress-mobile/WordPress-iOS/compare/task/22473-share-own-published-post...task/22473-share-own-published-post-hassaan-suggestion?expand=1). Let me know what you think!

**Comment by osullivanchris** [comment]
nice @hassaanelgarem I also noticed that area, but thought it might not be customisable. I also thought with my 'test post' title it might be less relatable than a real post title.

**Comment by hassaanelgarem** [comment]
> I implemented my suggestion for adding a message and pushed it [here](https://github.com/wordpress-mobile/WordPress-iOS/compare/task/22473-share-own-published-post...task/22473-share-own-published-post-hassaan-suggestion?expand=1). Let me know what you think!

Here's what it looks like for reference:

**Comment by salimbraksa** [comment]
> I implemented my suggestion for adding a message and pushed it here. Let me know what you think!

Nice improvement @hassaanelgarem, thanks for adding it. 🙇

**Comment by alpavanoglu** [comment]
Test case passes ✅ LGTM. 

1 minor thing could be doing the same thing we did for the avatars with `@ScaledMetric` for the action icon. Right now when the font size is increased, the icon looks rather small and might be hard to see. If you would like to apply this, you can pull from the feature branch and use the same value (`scale`) I added to the `NotificationsTableViewCellContent` file/

**Review by alpavanoglu — APPROVED** [review]
